src/main/java/net/ripe/rpki/validator3/domain/RpkiObjectCleanupService.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    /**
+     * Marks all RPKI objects that are reachable from a trust anchor by following the entries in the manifests.
+     * Objects that are no longer reachable will be deleted after a configurable grace duration.
+     */
+    @Scheduled(initialDelay = 60_000, fixedDelayString = "${rpki.validator.rpki.object.cleanup.interval.ms}")
+    public long cleanupRpkiObjects() {
+        Instant now = Instant.now();
+        for (TrustAnchor trustAnchor : trustAnchors.findAll()) {
+            transactionTemplate.execute((status) -> {
+                entityManager.setFlushMode(FlushModeType.COMMIT);
+
+                log.info("tracing objects for trust anchor {}", trustAnchor);
+
+                X509ResourceCertificate resourceCertificate = trustAnchor.getCertificate();
+                if (resourceCertificate != null) {
+                    traceCertificateAuthority(now, resourceCertificate);
+                }
+
+                return null;
+            });
+        }
+
+        return deleteUnreachableObjects(now);
+    }
+
+    private long deleteUnreachableObjects(Instant now) {
+        return transactionTemplate.execute((status) -> {
+            entityManager.flush();
+
+            Instant unreachableSince = now.minus(cleanupGraceDuration);
+            long count = rpkiObjects.deleteUnreachableObjects(unreachableSince);
+            log.info("Deleted {} RPKI objects that have not been marked reachable since {}", count, unreachableSince);
+            return count;
+        });
+    }
+
+    private void traceCertificateAuthority(Instant now, X509ResourceCertificate resourceCertificate) {
+        if (resourceCertificate == null || resourceCertificate.getManifestUri() == null) {
+            return;
+        }
+
+        Optional<RpkiObject> maybeManifest = rpkiObjects.findLatestByTypeAndAuthorityKeyIdentifier(
+            RpkiObject.Type.MFT,
+            resourceCertificate.getSubjectKeyIdentifier()
+        );
+        maybeManifest.ifPresent(manifest -> {
+            markAndTraceObject(now, "manifest.mft", manifest);
+        });
+    }
+
+    private void markAndTraceObject(Instant now, String name, RpkiObject rpkiObject) {
+        if (now == rpkiObject.getLastMarkedReachableAt()) {
+            log.warn("object already marked, skipping {}", rpkiObject);
+        }
+
+        rpkiObject.markReachable(now);
+        switch (rpkiObject.getType()) {
+            case MFT:
+                traceManifest(now, name, rpkiObject);
+                break;
+            case CER:
+                traceCaCertificate(now, name, rpkiObject);
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void traceManifest(Instant now, String name, RpkiObject manifest) {
+        manifest.get(ManifestCms.class, name).ifPresent(manifestCms -> {
+            manifestCms.getFiles().forEach((entry, sha256) -> {
+                rpkiObjects.findBySha256(sha256).ifPresent(rpkiObject -> {
```

Comment:
Shouldn't we better have rpkiObjects.findBySha256(listOfSha256s) method?
It's not that important, but as soon as we'll have a lot of object to mark, it could speed it up and avoid potential locks for other transactions (we had similar problem in current validator).

Reply:
This is definitely a `select 1+n` case. Not sure if querying a lists of objects will be that much faster (DB is local anyway). 

But sorting the hashes and then `SELECT FOR UPDATE` may reduce locking problems improve, so I'll add a `findObjectsInManifest` method.